### PR TITLE
homework.lambda.part2.exercise2 (yushina)

### PR DESCRIPTION
**Pull request by AnyaYu**

### CODE REVIEW

**Review by bormariya — APPROVED** [review]

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    //  - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
+    //  - при реализации использовать логический оператор &&
+    private Predicate<Person> andUsingLogicalOperator(Predicate<Person> left, Predicate<Person> right) {
+        return person -> left.test(person) && right.test(person);
+    }
+
+    @Test
+    void personHasNotEmptyLastNameAndFirstNameUsingLogicalOperators() {
+        Predicate<Person> personHasEmptyFirstName = person -> person.getFirstName().isEmpty() || person.getFirstName() == null;
+        Predicate<Person> personHasEmptyLastName = person -> person.getLastName().isEmpty() || person.getFirstName() == null;
```
Comment (milkplusplus) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: Mistake:
Predicate<Person> personHasEmptyLastName = person -> person.getLastName().isEmpty() || person.**getFirstName**() == null;

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+        Predicate<Person> personHasNotEmptyFirstName = negate(personHasEmptyFirstName);
```
Comment (milkplusplus) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: You need to use `negateUsingLogicalOperator(...)` here

File: src/test/java/lambda/part2/exercise/Exercise2.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    // - возвращает новый предикат, объединяющий исходные с помощью операции "AND"
+    // - при реализации использовать логический оператор &&
+    private <T> Predicate<T> and(Predicate<T> left, Predicate<T> right) {
+        return t -> left.test(t) && right.test(t);
+    }
+
+    @Test
+    void personHasNotEmptyLastNameAndFirstNameUsingGenericPredicates() {
+        Predicate<Person> personHasEmptyFirstName = person -> person.getFirstName().isEmpty() || person.getFirstName() == null;
+        Predicate<Person> personHasEmptyLastName = person -> person.getLastName().isEmpty() || person.getFirstName() == null;
```
Comment (milkplusplus) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: Again about `getFirstName()` in personHasEmptyLastName

**Review by milkplusplus — CHANGES_REQUESTED** [review]

File: src/test/java/lambda/part2/exercise/Exercise2.java (hunk 1 above)
Comment (AnyaYu) [reply src/test/java/lambda/part2/exercise/Exercise2.java]: Done

File: src/test/java/lambda/part2/exercise/Exercise2.java (hunk 2 above)
Comment (AnyaYu) [reply src/test/java/lambda/part2/exercise/Exercise2.java]: Done

File: src/test/java/lambda/part2/exercise/Exercise2.java (hunk 3 above)
Comment (AnyaYu) [reply src/test/java/lambda/part2/exercise/Exercise2.java]: Done

File: src/test/java/lambda/part2/exercise/Exercise2.java (hunk 1 above)
Comment (milkplusplus) [inline src/test/java/lambda/part2/exercise/Exercise2.java]: It's still here :)

Comment (AnyaYu) [reply src/test/java/lambda/part2/exercise/Exercise2.java]: ok now fixed

**Review by milkplusplus — APPROVED** [review]

**Pull request merged**